### DISCUSSION TITLE
Guidelines for plugins

### DISCUSSION THREAD

**Discussion by 6543** [woodpecker-ci/woodpecker#3085] [General]
Based on the idea of https://github.com/woodpecker-ci/woodpecker/pull/3062#issuecomment-1872563558

why not define some baseline criteria who plugins have to meet before we let them be added. this way we ensure the listed plugins have a certant quality.

like they must follow semantic-versioning or have a mechanism to automatically get notified or better also automatically get there dependency  updated ...

**Comment by pat-s** [comment]
Thanks for this discussion. 

I think guidelines will help both users and devs. I'll propose some points as a starting point for a discussion. Of course, these rules can only be enforced in forge orgs of the orgs, not outside. Putting a plugin into an org should come with some benefits. For example, I would suggest to only list plugins in the official overview if they live within the org and comply to these rules. Of course anyone can create any plugin (which is great) but the more plugins exists which follow the same rules, the easier it will be for users.


- **Versioning**: Use semver (discussion: with or without `v` prefix? We're not consistent here right now and I think it would help everyone if we would)
- **Maintainers**: Add a section to the README who maintains this plugin and should be tagged for PRs (this means we need to find official maintainers for every plugin but this will help everyone in the end I guess)
- **Dependencies**: Keep dependencies up-to-date. `renovate` is an option/offer, not a must. But it's easy to set up and if not used, we should define time-based rules how often dep updates should happen
- **Linters**: add linters (list: tdb). Precommit also helps but is optional.
- **PRs**: add workflow to build PR images for simplified testing
- **Release helper**: optional but encouraged. https://github.com/woodpecker-ci/plugin-ready-release-go - also soon available for Gitea
- **Forge choice**: add an information that plugins can either be added to CB or GH (whatever forge maintainers prefer). Maybe we should even open a dedicated `woodpecker-plugins` org on GH to have similar names and a dedicated place for plugins?

Let's discuss. I can edit/maintain the points above so we have a curated list and an overview of the discussion.

@woodpecker-ci/maintainers

**Reply by qwerty287** [reply]
I think, we need to make clear about what we're talking, because for me it looks like @6543 proposed guidelines for adding plugins to the index, but @pat-s is talking about adding plugins to the `woodpecker-plugins` org.

Let me start with a general question: Should we add externally developed (i.e. not by us) plugins to `woodpecker-plugins`?

I don't like this idea.

Until now, nothing like this happened, and I also don't really see a benefit if we allow/do it.
We have our index which is totally open to any new plugin (if it complains with the guidelines if we add any), and this index is and should be the central place for plugins, not an org. So why don't we just let users choose where to host their code (which forge), but under their own user accounts? This also has the advantage that you can directly see who develops this plugin (the repo owner) - using a central org does not allow this and you need to look at README first.

> For example, I would suggest to only list plugins in the official overview if they live within the org and comply to these rules.

As my comments above probably already made clear, I'm against this. I think it's better to allow the devs to choose where they host their code. What if I want to use a dedicated gitea instance?

> add linters (list: tdb). Precommit also helps but is optional.

While linters are good, but for simple plugins it's easy to ensure code quality by just looking at it.
Also, don't forget that plugins can use any stack, so we can't require a linter list (except dockerfile linting) as plugins can use any programming language but we can't have a list for every one.

I see your idea in having guidelines for plugins - and I also think it can be helpful -, but to be honest I currently don't really see a way to merge my own opinions with guidelines as I don't really want to have any "hard" (i.e. rules that depend on programming practices that can be different) guidelines for the plugin index.
The only idea I had is some kind of a "badge" (similar to the "verified" checkmark) which allows users to see whether a plugin follows certain standards.

However, we would need checks to find out if plugins follow the rules. We aren't able to review all plugins every few weeks to check if they follow the guidelines. I don't really see a good way to enforce the guidelines without just checking once on addition.

Just like always,these are just opinions of course 🙂

**Reply by pat-s** [reply]
> I think, we need to make clear about what we're talking, because for me it looks like @6543 proposed guidelines for adding plugins to the index, but @pat-s is talking about adding plugins to the woodpecker-plugins org.

My comment was targeting plugins within our orgs, yes, but not necessarily enforcing all plugins to be there. We can't do this/control this, of course. Yet, I strongly believe that there should be guidelines within our org for plugins (which is what this discussion is mainly about). I was missing them when I started "cleaning" them a few weeks/months ago and noticed that many of them differ in shared points (workflows, linters, update practices, README, etc.). It was more like an "oh this plugin has seen some (more) love 3 months ago, another one has been abandoned in categories X and Y for years but at least got a few bug fix releases in the code)."

Which is why I tried to harmonize them using best efforts of linters, workflows, updating rules, etc.
It's still not great IMO, and we recently had new users who were not sure why some "features" existed in some repos (e.g. the ability to build PR images) or some linter YAML files but were missing in others. And who is actually responsible, why are there two orgs, etc.

These are all valid user questions and easy to overlook from a maintainer's perspective/one which is with the project a little longer. They also don't look "clean and professional" to the outside.

> Should we add externally developed (i.e. not by us) plugins to woodpecker-plugins?

I think I was partly unclear. **If**, then they should apply to the (non-existing) rules yet and also  be actively maintained by that person. We have some recent examples here:

- https://github.com/woodpecker-ci/plugin-extend-env was started by @anbraten and @qwerty287 but initially requested by a user of the buildx plugin which then opened a discussion in the main repo and at some point wanted to create his own plugin
- https://codeberg.org/woodpecker-plugins/go-plugin-reviewdog is a new one started recently by lafriks
- [A user of the `node-pm` plugin wanted to create a new one](https://codeberg.org/woodpecker-plugins/node-pm/pulls/18#issuecomment-1397715) for publishing to "pages" recently and asked if it could be included in the org and if there are some dev guidelines (earlier on)

> As my comments above probably already made clear, I'm against this. I think it's better to allow the devs to choose where they host their code. What if I want to use a dedicated gitea instance?

Of course, it's their choice. But we shouldn't blindly list all plugins on the website if we can't (partly) guarantee for their correctness and safety. And ATM we already offer two forges for hosting. I think there should be a way for externals/non-maintainers to develop their plugin following org rules to gain visibility and user trust.

**Reply by qwerty287** [reply]
> My comment was targeting plugins within our orgs, yes, but not necessarily enforcing all plugins to be there. We can't do this/control this, of course. Yet, I strongly believe that there should be guidelines within our org for plugins (which is what this discussion is mainly about). I was missing them when I started "cleaning" them a few weeks/months ago and noticed that many of them differ in shared points (workflows, linters, update practices, README, etc.). It was more like an "oh this plugin has seen some (more) love 3 months ago, another one has been abandoned in categories X and Y for years but at least got a few bug fix releases in the code)."

So you would only create guidelines for plugins developed by us? (If we don't allow externally developed plugins.)

If yes, I finally got your points I think, and I also think it's fine to have rules there if we don't add 3rd-party plugins to our org.

> why are there two orgs

Yes, that's also an issue, but it's a different topic and should be discussed in #2143.

> also be actively maintained by that person

Well, then we don't have to add it to our org. The dev can just leave it under their own user account.

If we only allow plugins that follow specific rules, we would have to check if they still follow these rules and we don't have the resources for this.

> https://github.com/woodpecker-ci/plugin-extend-env was started by @anbraten and @qwerty287 but initially requested by a user of the buildx plugin which then opened a discussion in the main repo and at some point wanted to create his own plugin

Yes, it was requested, but we started development, so it's "our" plugin and we maintain it.

> [A user of the node-pm plugin wanted to create a new one](https://codeberg.org/woodpecker-plugins/node-pm/pulls/18#issuecomment-1397715) for publishing to "pages" recently and asked if it could be included in the org and if there are some dev guidelines (earlier on)

I don't think it's a good idea to let users create plugins and then just add them to the org because the users don't want to maintain it and we should do this. Of course, if we use the plugin in our own repos, we can do this, but otherwise I wouldn't allow this.

> I think there should be a way for externals/non-maintainers to develop their plugin following org rules to gain visibility and user trust.

Of course, but as I already wrote: We can't control this. We don't have the resources to check if plugins follow our standards.

Because long comments like these we're writing here are a bit unsuitable as overview, let me sum up my points again:
I don't think we should add external plugins to our org (`woodpecker-plugins`). If only plugins developed/maintained by us are there, I also think guidelines are a good idea then.
Adding an external plugin to the org should only happen if we as maintainers maintain the plugin then which should only happen if we see a big problem with a plugin e.g. not maintained anymore, but external developers shouldn't have write or even admin access to repos in the orgs, because I don't see a benefit in it. If we only allow plugins that follow specific guidelines - so users can easily see whether a plugin follows them and is trustable - we need to check periodically if *all* plugins in the org follow these, even if we as maintainers don't use and develop it, and we don't have the resources to do this. If we allow any plugin to be added to our org without having guidelines, it's useless anyways.
About the hoster of the org, we should move this discussion to #2143 or a new discussion, but it's not important for this issue where the code is hosted.

**Reply by pat-s** [reply]
Yes, long discussion. Usually, such discussions are tackled in personal meetings/calls. But here it's different, of course :)
Some clarifications, as I think my positions haven't' yet fully been understood:

> So you would only create guidelines for plugins developed by us? (If we don't allow externally developed plugins.)

They *should* apply to all but as we all know, we can only control/enforce them among our orgs. But if we encourage the development of external plugins, there should be a way to distinguish "official" plugins (which would imply following the developed rules) from others.

> Well, then we don't have to add it to our org. The dev can just leave it under their own user account.

There are big differences WRT to where the plugin lives. In our org we have the "control" (i.e. we can archive it if necessary, apply changes). In external repos, we don't.

> If we only allow plugins that follow specific rules, we would have to check if they still follow these rules and we don't have the resources for this

Yes, but we also don't want to have a "wild west" and just let everybody do whatever. This helps nobody. And by ensuring central workflows, linters, etc. a big chunk of this can be automated.

> I don't think it's a good idea to let users create plugins and then just add them to the org because the users don't want to maintain it and we should do this. Of course, if we use the plugin in our own repos, we can do this, but otherwise I wouldn't allow this.

Of course, that's a potential issue. That's why I requested *explicit* maintainers and not just "the maintainers team". As with the latter, nobody will do it as it's "only" a group. By having explicit maintainers, people can be "judged" and actions can be taken by admins if nothing happens. E.g. if an external person moves it to the org and then doesn't maintain it, we can archive it.

---

Overall, I see the biggest discrepancy in our view on the ability to enforce general rules on all repos and "trust" individual people to do their job WRT maintenance. As a user I would prefer seeing one or two dedicated people listed as being the responsible people to tag for an issue/PR instead of a large group and then "hope for the best". 

I've contributed many years to a similar large project which was split up into many small repos (similar to the plugin style here) and we had "great" success in moving away from "somebody of the group will surely eventually do it/reply" to assigning maintenance explicitly to individual users who make sure to enforce shared guidelines and reply to issues/PRs. (Happy to share more in private if of interest).

**Reply by qwerty287** [reply]
> Usually, such discussions are tackled in personal meetings/calls

Maybe it's time for https://github.com/woodpecker-ci/woodpecker/discussions/2146 :)

> As a user I would prefer seeing one or two dedicated people listed as being the responsible people to tag for an issue/PR instead of a large group and then "hope for the best".

Sure, that's a good idea, but it's easy to do this by just adding a comment to the readme. However, plugins having this should not be marked as "developed by Woodpecker Authors" anymore I think.

I think one of my issues is that I would like to separate official plugins - a different "official" than what you wrote, because I think "official" means developed by Woodpecker Authors - from 3rd-party plugins.

Maybe, your approach in having a central org with guidelines is a good idea, but then I'd like to suggest to have really official plugins in another org (e.g. the woodpecker main org). Of course, these should also follow any guidelines we introduce.
I'd say the plugins that are really developed by woodpecker authors and not just single persons of them are `plugin-git`, `plugin-docker-buildx`, and `plugin-ready-release-go`. Also, I'd add `plugin-github-release`, `plugin-surge-preview`, `plugin-prettier` and `plugin-codecov` to this list as they're also developed by multiple persons of us and they're pretty essential for our own workflows.

**Reply by zc-devs** [reply]
If I'm concerned, I'm with @qwerty287 mostly.

Make the rules, follow them, show best practices (let's call it).

However, I don't see the point to move third party plugins to the org. [The current list of plugins](https://woodpecker-ci.org/plugins) is enough to me. If I, as a third party plugin developer, want some respect/promotion in the list, then I can apply for the `trusted` label. In this case I should implement some rules, of course.

**Reply by pat-s** [reply]
I think I am still misunderstood. I didn't mean to say that plugins *must* be moved to the plugins org. We should *allow* people to move plugins there and maintain them (for centralization reasons) which would of course then come with the "burden" to apply to our "rules". 
The rule suggestions from above would, of course, only apply to the plugins within the org as we can only enforce them there.

If we add plugins to the plugins list, there must be a clear way to identify plugins which follow our rules/a certain standard compared to "indy" ones. Otherwise, users and the project itself suffers from potentially outdated or badly developed ones.

(@zc-devs  Not sure if I was really misunderstood but the part of your reply which says "However, I don't see the point to move third party plugins to the org" gives me the feeling that you understood it as me proposing "plugins **must** be moved to the org")

**Reply by zc-devs** [reply]
> allow ... burden

I got it. Don't want to repeat `qwerty`. That burden lean on maintainers also.

> for centralization reasons

There should be some "marketplace" of plugins, of course. To me, current list seems centralized and works as marketplace.

> identify plugins which follow our rules

Labels already there.

> potentially outdated or badly developed

Life is life. This is subjective. Everybody have own head.
In case of `trusted` plugins, maintainers just need to remove the label and not to deal with removing repository, taking down access, etc.
